File: src/domain/code.test.ts
import { describe, expect, it } from 'vitest';
import { levenshtein, normalizeCode, parseCode, toDisplay } from './code';

describe('normalizeCode', () => {
  it('uppercases and strips non-alphanumerics', () => {
    expect(normalizeCode('civ 12')).toBe('CIV12');
    expect(normalizeCode('c1v-12')).toBe('C1V12');
    expect(normalizeCode('  fwc_1  ')).toBe('FWC1');
    expect(normalizeCode('00')).toBe('00');
  });

  it('returns empty string for pure junk', () => {
    expect(normalizeCode('--- .')).toBe('');
  });
});

describe('toDisplay', () => {
  it('inserts a space between letters and digits', () => {
    expect(toDisplay('CIV12')).toBe('CIV 12');
    expect(toDisplay('FWC1')).toBe('FWC 1');
  });

  it('leaves all-digit or all-letter codes unchanged', () => {
    expect(toDisplay('00')).toBe('00');
    expect(toDisplay('FWC')).toBe('FWC');
  });

  it('leaves mixed/irregular codes unchanged', () => {
    expect(toDisplay('C1V12')).toBe('C1V12');
  });
});

describe('parseCode', () => {
  it('splits a letter prefix and trailing number', () => {
    expect(parseCode('CIV12')).toEqual({ teamCode: 'CIV', number: 12, canonical: 'CIV12' });
    expect(parseCode('civ 12')).toEqual({ teamCode: 'CIV', number: 12, canonical: 'CIV12' });
    expect(parseCode('FWC1')).toEqual({ teamCode: 'FWC', number: 1, canonical: 'FWC1' });
  });

  it('handles the pure-number logo sticker', () => {
    expect(parseCode('00')).toEqual({ teamCode: '', number: 0, canonical: '00' });
  });

  it('returns null when there are no trailing digits', () => {
    expect(parseCode('CIV')).toBeNull();
    expect(parseCode('')).toBeNull();
    expect(parseCode('AB12C')).toBeNull();
  });
});

describe('levenshtein', () => {
  it('is zero for identical strings', () => {
    expect(levenshtein('CIV12', 'CIV12')).toBe(0);
  });

  it('counts single edits', () => {
    expect(levenshtein('C1V12', 'CIV12')).toBe(1); // substitution
    expect(levenshtein('CIV2', 'CIV12')).toBe(1); // insertion
    expect(levenshtein('CIV123', 'CIV12')).toBe(1); // deletion
  });

  it('handles empty strings', () => {
    expect(levenshtein('', '')).toBe(0);
    expect(levenshtein('', 'ABC')).toBe(3);
    expect(levenshtein('ABC', '')).toBe(3);
  });

  it('is symmetric', () => {
    expect(levenshtein('EGYA', 'EGY4')).toBe(levenshtein('EGY4', 'EGYA'));
  });

  it('counts multiple edits', () => {
    expect(levenshtein('ZZZ99', 'CIV12')).toBe(5);
  });
});

File: android/app/src/test/java/br/com/fiquemsabendo/figurinhas/domain/CodeTest.kt
package br.com.fiquemsabendo.figurinhas.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class CodeTest {
    @Test fun normalize_uppercases_and_strips() {
        assertEquals("CIV12", normalizeCode("civ 12"))
        assertEquals("C1V12", normalizeCode("c1v-12"))
        assertEquals("FWC1", normalizeCode("  fwc_1  "))
        assertEquals("00", normalizeCode("00"))
    }

    @Test fun normalize_empty_for_junk() {
        assertEquals("", normalizeCode("--- ."))
    }

    @Test fun toDisplay_inserts_space() {
        assertEquals("CIV 12", toDisplay("CIV12"))
        assertEquals("FWC 1", toDisplay("FWC1"))
    }

    @Test fun toDisplay_leaves_pure_and_irregular() {
        assertEquals("00", toDisplay("00"))
        assertEquals("FWC", toDisplay("FWC"))
        assertEquals("C1V12", toDisplay("C1V12"))
    }

    @Test fun parse_splits_prefix_and_number() {
        assertEquals(ParsedCode("CIV", 12, "CIV12"), parseCode("CIV12"))
        assertEquals(ParsedCode("CIV", 12, "CIV12"), parseCode("civ 12"))
        assertEquals(ParsedCode("FWC", 1, "FWC1"), parseCode("FWC1"))
    }

    @Test fun parse_pure_number_logo() {
        assertEquals(ParsedCode("", 0, "00"), parseCode("00"))
    }

    @Test fun parse_null_without_trailing_digits() {
        assertNull(parseCode("CIV"))
        assertNull(parseCode(""))
        assertNull(parseCode("AB12C"))
    }

    @Test fun levenshtein_basics() {
        assertEquals(0, levenshtein("CIV12", "CIV12"))
        assertEquals(1, levenshtein("C1V12", "CIV12"))   // substitution
        assertEquals(1, levenshtein("CIV2", "CIV12"))    // insertion
        assertEquals(1, levenshtein("CIV123", "CIV12"))  // deletion
        assertEquals(0, levenshtein("", ""))
        assertEquals(3, levenshtein("", "ABC"))
        assertEquals(3, levenshtein("ABC", ""))
        assertEquals(levenshtein("EGY4", "EGYA"), levenshtein("EGYA", "EGY4"))
        assertEquals(5, levenshtein("ZZZ99", "CIV12"))
    }
}
